### PR TITLE
Updated Distribution Contract

### PR DESCRIPTION
**Pull request by donchate**
Complete rewrite of the distribution contract to add a new distribution strategy using diesel pool liquidity providers and retain existing functionality for fixed distributions.

Both strategies now distribute by virtual transaction over a specified number of ticks (similar to unstaking), or the distribution owner can flush the pending amounts manually. Initial tick time is set to 24 hours for a daily distribution.

~~WIP: payout tx limit and spreading over future blocks~~

depends on #81

### CODE REVIEW

File: contracts/distribution.js [hunk 1]
```diff
@@ -203,10 +198,25 @@ actions.update = async (payload) => {
   if (api.assert(authorizedCreation, 'you must have enough tokens to cover the update fee')
     && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')) {
     const exDist = await api.db.findOne('batches', { _id: id });
-    if (api.assert(exDist, 'distribution not found') && await validateMinPayout(tokenMinPayout)
-      && await validateRecipients(tokenRecipients)) {
-      exDist.tokenMinPayout = tokenMinPayout;
-      exDist.tokenRecipients = tokenRecipients;
+    if (api.assert(exDist, 'distribution not found')) {
+      if (numTicks && api.assert(typeof numTicks === 'string' && api.BigNumber(numTicks).gt(0) && api.BigNumber(numTicks).lte(5555), 'numTicks must be a number between 1 and 5555')) {
+        exDist.numTicks = numTicks;
```
Comment (bt-cryptomancer) [inline contracts/distribution.js:203]: If numTicks is updated, should numTicksLeft be updated as well?

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
Mostly looks good.  Aside from the comments I left about numTicks, please consider the following:

1.  I see you added some new fields to the data in the batches table.  What is your plan for handling old records in the batches table that pre-date this change?  One possible method would be to add code in the createSSC action that migrates existing batches records to the new format (adds the new fields); the botcontroller contract has an example of this technique.  Another strategy is to make it backwards compatible in case the new fields don't exist for a given record.

2.  Please remove the changes to libs/Block.js.  We'll have the checkPendingDistributions virtual tx be the first use of configurable virtual transactions via @eonwarped 's work in https://github.com/hive-engine/steemsmartcontracts/pull/81 .  You may have to merge his branch into your own in order to get your test cases to work.  You can refer to the tests in his PR to see how checkPendingDistributions should be configured.

3.  You mentioned "WIP: payout tx limit and spreading over future blocks".  I think this is definitely important and should be in place before releasing this PR.  Let me know when this chunk is ready for review.

File: contracts/distribution.js (hunk 1 above)
Comment (donchate) [reply contracts/distribution.js:215]: Yes it should. Updating numTicks will always restart the payment schedule.

File: contracts/distribution.js [hunk 2]
```diff
@@ -203,10 +198,25 @@ actions.update = async (payload) => {
   if (api.assert(authorizedCreation, 'you must have enough tokens to cover the update fee')
     && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')) {
     const exDist = await api.db.findOne('batches', { _id: id });
-    if (api.assert(exDist, 'distribution not found') && await validateMinPayout(tokenMinPayout)
-      && await validateRecipients(tokenRecipients)) {
-      exDist.tokenMinPayout = tokenMinPayout;
-      exDist.tokenRecipients = tokenRecipients;
+    if (api.assert(exDist, 'distribution not found')) {
+      if (numTicks && api.assert(typeof numTicks === 'string' && api.BigNumber(numTicks).gt(0) && api.BigNumber(numTicks).lte(5555), 'numTicks must be a number between 1 and 5555')) {
```
[the remark this answers is not in the document]
Comment (donchate) [reply contracts/distribution.js]: Added

File: contracts/distribution.js [hunk 3]
```diff
@@ -146,47 +140,48 @@ actions.create = async (payload) => {
...
-  } = payload;
-
-  const dist = await api.db.findOne('batches', { _id: id });
-  if (api.assert(dist, 'distribution id not found')
     && api.assert(isSignedWithActiveKey === true, 'you must use a transaction signed with your active key')
-    && api.assert(api.sender === api.owner || api.sender === dist.creator, 'must be contract owner or creator')) {
-    if (await processBatch(dist, symbol, true)) {
-      api.emit('flush', { memo: `${symbol} payout distributed` });
+    && api.assert(typeof strategy === 'string' && DistStrategy.indexOf(strategy) !== -1, 'invalid strategy')
+    && api.assert(typeof numTicks === 'string' && api.BigNumber(numTicks).gt(0) && api.BigNumber(numTicks).lte(5555), 'numTicks must be a number between 1 and 5555')) {
```
[the remark this answers is not in the document]
Comment (donchate) [reply contracts/distribution.js]: Added

**Comment by donchate** [comment]
Updated to use `registerTick` and limit payments to a default param of 50 per block. Overflow payments are stored in DB and attempted to pay every block that isn't already processing a distribution.

> 1. I see you added some new fields to the data in the batches table.  What is your plan for handling old records in the batches table that pre-date this change?  One possible method would be to add code in the createSSC action that migrates existing batches records to the new format (adds the new fields); the botcontroller contract has an example of this technique.  Another strategy is to make it backwards compatible in case the new fields don't exist for a given record.

There haven't been any distribution batches created to worry about yet.

**Comment by donchate** [comment]
The final commit adds `timeFactor` to liquidity positions to track how long they've had their position in place. This is to ensure liquidity is in place for a minimum time period before being eligible for reward, as well as provides data for a future revision that could bonus longer providers.

**Review by bt-cryptomancer — CHANGES_REQUESTED** [review]
Mostly looks good, adjustments for earlier feedback are fine, just want to check if you've considered the additional things I've pointed out in the comments.

File: contracts/marketpools.js [hunk 4]
```diff
@@ -335,15 +361,24 @@ actions.addLiquidity = async (payload) => {
...
     if (lp) {
-      lp.shares = api.BigNumber(lp.shares).plus(newShares);
+      const existingShares = lp.shares;
+      const finalShares = api.BigNumber(lp.shares).plus(newShares);
+      const timeOffset = api.BigNumber(finalShares).minus(existingShares).abs().dividedBy(existingShares);
+      lp.shares = finalShares;
+      lp.timeFactor = api.BigNumber.min(
+        api.BigNumber(lp.timeFactor).times(api.BigNumber('1').plus(timeOffset)),
+        blockDate.getTime(),
+      ).toNumber();
```
[the remark this answers is not in the document]
Comment (donchate) [reply contracts/marketpools.js:374]: Yes it should. Added rounding.

File: contracts/marketpools.js [hunk 5]
```diff
@@ -335,15 +361,24 @@ actions.addLiquidity = async (payload) => {
     if (!api.assert(api.BigNumber(newShares).gt(0), 'insufficient liquidity created')) return;
 
     // update liquidity position
+    const blockDate = new Date(`${api.hiveBlockTimestamp}.000Z`);
     const lp = await api.db.findOne('liquidityPositions', { account: api.sender, tokenPair });
     if (lp) {
-      lp.shares = api.BigNumber(lp.shares).plus(newShares);
+      const existingShares = lp.shares;
+      const finalShares = api.BigNumber(lp.shares).plus(newShares);
+      const timeOffset = api.BigNumber(finalShares).minus(existingShares).abs().dividedBy(existingShares);
```
[the remark this answers is not in the document]
Comment (donchate) [reply contracts/marketpools.js:369]: timeOffset is used once as an intermediate value, rounding has been added to the output value that is saved. I think in the intermediate state as a BigNumber obj should be fine.

File: contracts/distribution.js [hunk 6]
```diff
@@ -244,43 +268,224 @@ actions.deposit = async (payload) => {
...
+          payResult = await payRecipient(tr.account, payTokens[i].symbol, payoutQty);
+        } else {
+          const storePending = {
+            distId: dist._id,
+            account: tr.account,
+            symbol: payTokens[i].symbol,
+            quantity: payoutQty,
+            dueTime: blockDate.getTime(),
+          };
+          payResult = await api.db.insert('pendingPayments', storePending);
```
[the remark this answers is not in the document]
Comment (donchate) [reply contracts/distribution.js]: A quick test indicated it far faster to insert one row with all of the pending data so I went with that. Values below are for 1,000,000 sample records.
Some averages from a non-SSD test system:
Inserting in loop 731ms
Inserting one row <1ms

[the remark this answers is not in the document]
Comment (bt-cryptomancer) [reply contracts/distribution.js]: Okay perfect.

File: contracts/marketpools.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (bt-cryptomancer) [reply contracts/marketpools.js:369]: okay, sounds good

**Review by bt-cryptomancer — DISMISSED** [review]
Everything looks good now, approved!

**Review by bt-cryptomancer — APPROVED** [review]
changes look good

**Pull request merged**